milli/src/update/word_prefix_pair_proximity_docids.rs

```diff
@@ -1,16 +1,210 @@
+database before proceeding any further. This operation is called
+“flushing the batch”. Flushing the batch should also be done whenever `word1`
+is different than the previous `word1`.
+
+6. **Flushing the batch:** to flush the batch, we look at the `word1` and
+iterate over the elements of the batch in sorted order:
+```text
+Flushing Batch loop 1:
+------------------------------
+word1    : good
+word2    : d
+proximity: 1
+docids   : [docids2, docids3]
+```
+We then merge the array of `docids` (of type `Vec<Vec<u8>>`) using
+`merge_cbo_roaring_bitmap` in order to get a single byte vector representing a
+roaring bitmap of all the document ids where `word1` is followed by `prefix`
+at a distance of `proximity`.
+Once we have done that, we insert (`word1`, `prefix`, `proximity`) -> `merged_docids`
+into the database.
+
+7. That's it! ... except...
+
+## How is it created/updated (continued)
+
+I lied a little bit about the input data. In reality, we get two sets of the
+inputs described above, which come from different places:
+
+* For the list of sorted prefixes, we have:
+    1. `new_prefixes`, which are all the prefixes that were not present in the
+    database before the insertion of the new documents
+
+    2. `common_prefixes` which are the prefixes that are present both in the
+    database and in the newly added documents
+
+* For the list of word pairs and proximities, we have:
+    1. `new_word_pairs`, which is the list of word pairs and their proximities
+    present in the newly added documents
+
+    2. `word_pairs_db`, which is the list of word pairs from the database.
+    This list includes all elements in `new_word_pairs`** since `new_word_pairs`
+    was added to the database prior to calling the `WordPrefixPairProximityDocIds::execute`
+    function.
+
+To update the prefix database correctly, we call the algorithm described earlier first
+on (`common_prefixes`, `new_word_pairs`) and then on (`new_prefixes`, `word_pairs_db`).
+Thus:
+
+1. For all the word pairs that were already present in the DB, we insert them
+again with the `new_prefixes`. Calling the algorithm on them with the
+`common_prefixes` would not result in any new data.
+
+2. For all the new word pairs, we insert them twice: first with the `common_prefixes`,
+and then, because they are part of `word_pairs_db`, with the `new_prefixes`.
+
+Note, also, that since we read data from the database when iterating over
+`word_pairs_db`, we cannot insert the computed word-prefix-pair-proximity-
+docids from the batch directly into the database (we would have a concurrent
+reader and writer). Therefore, when calling the algorithm on
+(`new_prefixes`, `word_pairs_db`), we insert the computed
+((`word`, `prefix`, `proximity`), `docids`) elements in an intermediary grenad
+Writer instead of the DB. At the end of the outer loop, we finally read from
+the grenad and insert its elements in the database.
+
+
+
+*/
```

Comment:
sorry this is coming out of the blue, I may have missed the notification here for some reasons. anyway happy to discuss if we can make an article. Can we have a quick chat about it today @ManyTheFish @loiclec ? otherwise when I'll be back from vacations in 2 weeks?

Reply:
Hello @ferdi05, speaking with @loiclec, there are some improvements that can be done on this part, so, waiting until the end of your vacation seems to be better. 😄